test/xds/xds_client_outlier_detection_test.go

```diff
@@ -107,52 +109,91 @@ func clusterWithOutlierDetection(clusterName, edsServiceName string, secLevel e2
 		MaxEjectionPercent:             &wrapperspb.UInt32Value{Value: 1},
 		FailurePercentageThreshold:     &wrapperspb.UInt32Value{Value: 50},
 		EnforcingFailurePercentage:     &wrapperspb.UInt32Value{Value: 100},
-		FailurePercentageRequestVolume: &wrapperspb.UInt32Value{Value: 1},
-		FailurePercentageMinimumHosts:  &wrapperspb.UInt32Value{Value: 1},
+		FailurePercentageRequestVolume: &wrapperspb.UInt32Value{Value: 8},
+		FailurePercentageMinimumHosts:  &wrapperspb.UInt32Value{Value: 3},
 	}
 	return cluster
 }
 
+// checkRoundRobinRPCs verifies that EmptyCall RPCs on the given ClientConn,
+// connected to a server exposing the test.grpc_testing.TestService, are
+// roundrobined across the given backend addresses.
+//
+// Returns a non-nil error if context deadline expires before RPCs start to get
+// roundrobined across the given backends.
+func checkRoundRobinRPCs(ctx context.Context, client testpb.TestServiceClient, addrs []resolver.Address) error {
+	wantAddrCount := make(map[string]int)
+	for _, addr := range addrs {
+		wantAddrCount[addr.Addr]++
+	}
+	for ; ctx.Err() == nil; <-time.After(time.Millisecond) {
+		// Perform 3 iterations.
+		var iterations [][]string
+		for i := 0; i < 3; i++ {
+			iteration := make([]string, len(addrs))
+			for c := 0; c < len(addrs); c++ {
+				var peer peer.Peer
+				client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer))
+				if peer.Addr != nil {
+					iteration[c] = peer.Addr.String()
+				}
+			}
+			iterations = append(iterations, iteration)
+		}
+		// Ensure the the first iteration contains all addresses in addrs.
+		gotAddrCount := make(map[string]int)
+		for _, addr := range iterations[0] {
+			gotAddrCount[addr]++
+		}
+		if diff := cmp.Diff(gotAddrCount, wantAddrCount); diff != "" {
+			continue
+		}
+		// Ensure all three iterations contain the same addresses.
+		if !cmp.Equal(iterations[0], iterations[1]) || !cmp.Equal(iterations[0], iterations[2]) {
+			continue
+		}
+		return nil
+	}
+	return fmt.Errorf("timeout when waiting for roundrobin distribution of RPCs across addresses: %v", addrs)
+}
+
 // TestOutlierDetectionWithOutlier tests the Outlier Detection Balancer e2e. It
 // spins up three backends, one which consistently errors, and configures the
 // ClientConn using xDS to connect to all three of those backends. The Outlier
 // Detection Balancer should eject the connection to the backend which
-// constantly errors, and thus RPC's should mainly go to backend 1 and 2.
+// constantly errors, causing RPC's to not be routed to that upstream, and only
+// be Round Robined across the two healthy upstreams. Other than the intervals
+// the unhealthy upstream is ejected, RPC's should regularly round robin across
```

Comment:
` Other than the intervals the unhealthy upstream is ejected` ? Does this comment need updating?

Reply:
Sorry, I think this comment still applies. The upstream is ejected for only a certain amount of intervals, which is why for making this test deterministic I needed to only check round robin at some stage. I really appreciate your input and time spent reviewing, sorry I couldn't take any more of your changes.